02-intermediate-ruby/object-inheritance.md

```diff
@@ -3,77 +3,138 @@
 
-A way to describe this is by using the _is-a_ phrase: `Penguin` _is-a_ `Bird`. `Hawk` _is-a_ `Bird`. In Ruby, we use the `<` symbol to denote that the class on the left of the `<` symbol is inheriting behaviors from the class on the right side of the `<` symbol:
+A way to describe this is by using the _is-a_ phrase: a `Book` _is-a_ `Product`. `Video` _is-a_ `Product`. In Ruby, we use the `<` symbol to denote that the class on the left of the `<` symbol is inheriting behaviors from the class on the right side of the `<` symbol:
 
 ```ruby
-class Bird
-  def talk
-    "I'm a bird. Kah."
+class Product
+  attr_reader :id, :name
+
+  def initialize
+    @id = 12345
+    @name = "Generic Product"
+  end
+
+  def about
+    return "#{id}:  #{name}"
   end
 end
 
-# Penguin is-a Bird
-# Penguin inherits from Bird
-class Penguin < Bird
+# Book is-a Product
+# Book inherits from Product
+class Book < Product
 end
 ```
 
+The code above creates an is-a relationship between `Product` and `Book`.  The relationship diagramed below gives any instance of `Book` both `id`,  name and about methods.
+
+![Inheritance Diagram](images/basicInheritance.png)
+
 ## Inheriting Methods
-So what do we get when we inherit? We get __methods__. The  `Bird` class has an instance method called `talk`. Since all birds talk, our _subclasses_ will _inherit_ this method. So now our `Penguin` instances will have the talk method.
+So what do we get when we inherit? We get __methods__ and instance variables from the parent class. The  `Product` class has an instance method called `about`. Since all items in the store have an `about` method, our _subclasses_ will _inherit_ this method. So now our `Book` instances will have the `about` method.
 
 ```ruby
-generic_bird = Bird.new
-generic_bird.talk #=> "I'm a bird. Kah."
+generic_item = Product.new
+generic_item.about #=> "12345: Generic Product"
 
-tux = Penguin.new
-tux.talk # => "I'm a bird. Kah."
+poodr = Book.new
+poodr.about #=> "12345: Generic Product"
 ```
 
-This is known as _implicit inheritance_ because when we put functionality in the _base class_ (`Bird`), then all _subclasses_ (`Penguin`) will automatically get those behaviors.
+This is known as _implicit inheritance_ because when we put functionality in the _base class_ (`Product`), then all _subclasses_ (`Book`) will automatically get those behaviors.
 
 ## Overriding Methods
 What happens when our classes have different functionality for the same idea? We can _override_ methods from _base classes_ in their _subclasses_.
 
 ```ruby
-class Penguin < Bird
-  def talk
-    "I'm a bird. Sqwaaa."
+class Book < Product
+  def about
+    return "#{id}: Book"
   end
```

Comment:
The implementation of `Book#about` feels a little contrived to me. Is there a way we could provide a more intuitive example for these `Overriding Methods` and `Using super` sections?

Here is one idea that is pretty similar in complexity, but maybe a little more natural:

- Setup (previous section)
  - Each product knows what aisle it lives on
  - Product has a `find_in_store` method, that returns `"aisle #{@aisle_number}"`
- Overriding Methods
  - For a book, knowing the author's name is helpful in finding it on the shelf. We need to change the `find_in_store` method to give this information.
  - Now return `"aisle #{@aisle_number}, author #{author}"`
  - This also demonstrates that you can use instance variables from the parent in the child
- Using `super`
  - This isn't very DRY
  - Change it to `super + ", author #{author}"`

In order to make this work you'll need to move the bit on constructors up to before you talk about overriding methods. This is OK - it's a special case, and showing the pattern of "call `super` in the constructor" early is probably a good thing.

Let me know what you think.